gen-node-cfg.sh

```diff
@@ -0,0 +1,171 @@
+#!/bin/bash
+#Copyright 2017 William Stearns and Offensive Countermeasures
+
+
+#This and node.cfg-template should be in the same directory.
+#To test with different numbers of interfaces:
+#	sudo modprobe dummy numdummies=30
+#To remove these:
+#	sudo rmmod dummy
+#Once removed, one can rerun modprobe with a different number of dummies.
+#To see the available interfaces, run:
+#	sudo ip -o link
+
+require_file () {
+	#Returns true if all files or directories listed on the command line exist, False if one or more missing.
+
+	while [ -n "$1" ]; do
+		if [ ! -e "$1" ]; then
+			echo "Missing object $1. Please install it. Exiting." >&2
+			return 1					#False, at least one file or dir missing
+		fi
+		shift
+	done
+	return 0							#True, all objects are here
+}
+
+
+subtract_lists () {
+	#Returns all lines in the first list ("$1", one line per entry, surround by double quotes when passing in) that aren't in the second (same notes).
+
+	(
+		echo "${1}" | sort -u
+		echo "${2}"
+		echo "${2}"
+	) \
+	 | sort \
+	 | uniq -u
+}
+
+
+available_cores () {
+	#Returns the number of available processor cores.
+
+	grep '^processor\W\W*:\W[0-9]' /proc/cpuinfo | wc -l
+}
+
+
+available_interfaces () {
+	#Returns a list of all non-loopback interfaces, one per line
+	raw_if_list=`ip -o link | awk '{print $2}' | sed -e 's/:$//' | egrep -v '(^lo$)'`
+	default_ifs=`/sbin/ip route | grep '^default ' | sed -e 's/^.* dev //' -e 's/  *$//'`
+	non_default_ifs=`subtract_lists "$raw_if_list" "$default_ifs"`
+	echo "$non_default_ifs" | tr '\n' ' '
+}
+
+if_stats_for () {
+	#Returns the send/receive stats for the sole interface supplied as a parameter.
+	if [ -n "$1" ]; then
+		ip -s -o link | grep ' '"$1"':' | sed -e 's/\\/\n/g'		#Note; can't use "-h" for human readable as older versions of ip don't support it.
+	else
+		echo "No interface name supplied to if_stats_for."
+	fi
+}
+
+askYN () {
+	TESTYN=""
+	while [ "$TESTYN" != 'Y' ] && [ "$TESTYN" != 'N' ] ; do
+		echo -n '? ' >&2
+		read TESTYN || :
+		case $TESTYN in
+		T*|t*|Y*|y*)		TESTYN='Y'	;;
+		F*|f*|N*|n*)		TESTYN='N'	;;
+		esac
+	done
+
+	if [ "$TESTYN" = 'Y' ]; then
+		return 0 #True
+	else
+		return 1 #False
+	fi
+} #End of askYN
+
+
+#======== Main ========
+
+PATH="/bin:/sbin:/usr/bin:/usr/sbin:$PATH"
+export PATH
+
+require_file /bin/awk /bin/cp /bin/date /bin/egrep /bin/sed /bin/tr /proc/cpuinfo /sbin/ip || exit 1
```

Comment:
These tools can be at other locations. For instance, on my Ubuntu Server 16.04.3 system `awk` and `tr` are both in `/usr/bin`. Can we support if the tools are located anywhere in the user's path?

Reply:
Agreed - new version will look for these tools anywhere in the path.